File: haskell/lib/Ppc/Match.hs
module Match where

import Control.Applicative (Alternative ((<|>), empty), many, optional, some)
import Control.Monad (guard)
import Data.Char (isAlphaNum)
import Data.Functor (($>))
import Data.Maybe (catMaybes)
import GHC.Generics (Generic)
import Text.Parsec (Parsec, char, choice, many1, noneOf, optionMaybe, string, try, (<?>))
import qualified Text.Parsec as P

data Hole
  = Hole
      { pattern :: String,
        variable :: String,
        offset :: Integer,
        kind :: String
      }
  | Constant String
  deriving (Eq, Show, Generic)

-- Parser for identifiers
identifier :: Parsec String () String
identifier = (:) <$> p_identifier_first <*> many p_identifier_char
  where
    p_identifier_first = identStart
    p_identifier_char = identLetter <|> digit <|> charSingleQuote
    identStart = letter <|> charSingleQuote <|> charUnderscore
    identLetter = letter <|> charUnderscore
    charUnderscore = char '_'
    charSingleQuote = char '\''
    digit = noneOf ""

-- Parses the hole pattern with optional delimiters
holePattern :: String -> String -> Parsec String () (String, String)
holePattern left right = do
  l <- optional $ string left
  v <- identifier
  r <- optional $ string right
  return (l ++ v ++ r, v)

-- Parses the regex pattern with a separator
regexPattern :: String -> Char -> String -> Parsec String () (String, String)
regexPattern left sep right = do
  string left
  (v, expr) <- regexBody sep right
  string right
  return (v ++ [sep] ++ expr, v)

-- Parses the regex body
regexBody :: Char -> String -> Parsec String () (String, String)
regexBody sep suffix = do
  v <- identifier
  _ <- char sep
  expr <- regexExpression suffix
  return (v, expr)

-- Parses the regex expression
regexExpression :: String -> Parsec String () String
regexExpression suffix =
  concat <$> many1 (regexTerm suffix)

-- Parses a term in the regex expression
regexTerm :: String -> Parsec String () String
regexTerm suffix =
  charClass
    <|> escapedChar
    <|> ((:) <$> noneOf (suffix ++ "[]\\") <*> pure "") -- TODO: Fix this line
    <?> "regex term"
  where
    charClass = ("[" ++) <$> (char '[' *> regexExpression (']' : suffix))
    escapedChar = ('\\' :) <$> (char '\\' *> anyChar)

-- Parses an attribute access
attributeAccess :: Parsec String () String
attributeAccess =
  char '.'
    *> choice
      [ string "value",
        string "length",
        string "lines",
        string "offset.start",
        string "offset.end",
        string "offset",
        string "line.start",
        string "line.end",
        string "line",
        string "column.start",
        string "column.end",
        string "column",
        string "file.path",
        string "file.name",
        string "file.directory",
        string "file",
        string "lowercase",
        string "UPPERCASE",
        string "Capitalize",
        string "uncapitalize",
        string "UpperCamelCase",
        string "lowerCamelCase",
        string "UPPER_SNAKE_CASE",
        string "lower_snake_case",
        string "lsif.hover"
      ]

-- Parses a hole or constant
parseTemplate :: Parsec String () [Hole]
parseTemplate =
  catMaybes
    <$> many
      ( try (parseHole <?> "hole")
          <|> parseConstant
      )
  where
    parseHole = do
      offset <- P.getOffset
      (pattern, variable) <- choice holePatterns
      kind <- optional attributeAccess
      return $ Just $ Hole pattern variable (fromIntegral offset) (fromMaybe "value" kind)
    parseConstant =
      (Constant <$> many1 (noneOf ['$'])) `map` Just
    holePatterns =
      [ holePattern "" "" <?> "hole without delimiters",
        holePattern "[" "]" <?> "hole with square brackets",
        holePattern "{" "}" <?> "hole with curly braces",
        holePattern "(" ")" <?> "hole with parentheses",
        holePattern "<" ">" <?> "hole with angle brackets",
        regexPattern "$" '~' "$" <?> "regex hole"
      ]

-- Parses the given string into a template
parseTemplate' :: String -> Either ParseError [Hole]
parseTemplate' = P.parse parseTemplate ""

File: ocaml/lib/ppc/match.ml
module Match = struct
  open Parser

  type hole =
    | Hole of string * string * int * string
    | Constant of string

  let identifier =
    let identStart = word_sat (fun c -> Char.is_alpha c || c = '_' || c = '\'') in
    let identChar =
      word_sat
        (fun c -> Char.is_alpha c || Char.is_digit c || c = '_' || c = '\'')
    in
    identStart <*> (many identChar)

  let holePattern left right =
    let lp = option "" (string left) in
    let rp = option "" (string right) in
    let pat = lp <*> identifier <*> rp in
    pat >>| fun (l, v, r) -> (l ^ v ^ r, v)

  let regexPattern left sep right =
    string left
    >>= fun () ->
    let regexBody =
      identifier <*> (char sep <*> regexExpression (sep :: String.to_seq right))
    in
    regexBody >>= fun (v, expr) ->
    string right >>| fun () -> (v ^ String.make 1 sep ^ expr, v)

  and regexExpression suffix =
    let term = regexTerm (List.of_seq suffix) in
    many1 term >>| String.concat ""

  and regexTerm suffix =
    let charClass =
      char '['
      >>= fun () ->
      regexExpression (']' :: suffix) >>| fun expr -> "[" ^ expr
    in
    let escapedChar = char '\\' >>= fun () -> any_char >>| String.make 1 in
    charClass
    <|> escapedChar
    <|> (word_sat (fun c -> not (List.mem c suffix)) <|> return "")

  let attributeAccess =
    char '.'
    >>= fun () ->
    choice
      [
        "value";
        "length";
        "lines";
        "offset.start";
        "offset.end";
        "offset";
        "line.start";
        "line.end";
        "line";
        "column.start";
        "column.end";
        "column";
        "file.path";
        "file.name";
        "file.directory";
        "file";
        "lowercase";
        "UPPERCASE";
        "Capitalize";
        "uncapitalize";
        "UpperCamelCase";
        "lowerCamelCase";
        "UPPER_SNAKE_CASE";
        "lower_snake_case";
      ]

  let parseTemplate =
    let parseHole =
      get_offset
      >>= fun offset ->
      choice
        [
          holePattern "" "" >>| fun (pat, var) ->
            Hole (pat, var, offset, "value");
          holePattern "[" "]" >>| fun (pat, var) ->
            Hole (pat, var, offset, option "value" attributeAccess);
          holePattern "{" "}" >>| fun (pat, var) ->
            Hole (pat, var, offset, option "value" attributeAccess);
          holePattern "(" ")" >>| fun (pat, var) ->
            Hole (pat, var, offset, option "value" attributeAccess);
          holePattern "<" ">" >>| fun (pat, var) ->
            Hole (pat, var, offset, option "value" attributeAccess);
          regexPattern "$" '~' "$" >>| fun (pat, var) ->
            Hole (pat, var, offset, "value");
        ]
    in
    let parseConstant = word_sat (fun c -> c <> '$') >>| fun s -> Constant s in
    many (parseHole <|> parseConstant)

  let parseTemplate' input =
    let result = parse_string parseTemplate () input in
    match result with
    | Success (holes, _) -> Ok holes
    | Failed (error, _) -> Error error
end